src/back/msl/writer.rs

```diff
@@ -3901,6 +3881,218 @@ impl<W: Write> Writer<W> {
+            }
+
+            Ok(())
+        }
+
+        fn write_workgroup_variable_initialization(
+            &mut self,
+            module: &crate::Module,
+            ty: Handle<crate::Type>,
+            access_stack: &mut AccessStack,
+            level: back::Level,
+        ) -> BackendResult {
+            if does_ty_contain_atomic(module, ty)? {
+                match module.types[ty].inner {
+                    crate::TypeInner::Atomic { .. } => {
+                        write!(
+                            self.out,
+                            "{}{}::atomic_store_explicit({}",
+                            level, NAMESPACE, ATOMIC_REFERENCE
+                        )?;
+                        access_stack.write(&mut self.out, &self.names)?;
+                        writeln!(self.out, ", 0, {}::memory_order_relaxed);", NAMESPACE)?;
+                    }
+                    crate::TypeInner::Array { base, size, .. } => {
+                        let count = match size.to_indexable_length(module).expect("Bad array size")
+                        {
+                            proc::IndexableLength::Known(count) => count,
+                            proc::IndexableLength::Dynamic => return Ok(()),
+                        };
+                        for i in 0..count {
+                            access_stack.enter(Access::Index(i), |access_stack| {
+                                self.write_workgroup_variable_initialization(
+                                    module,
+                                    base,
+                                    access_stack,
+                                    level,
+                                )
+                            })?;
+                        }
+                    }
+                    crate::TypeInner::Struct { ref members, .. } => {
+                        for (index, member) in members.iter().enumerate() {
+                            access_stack.enter(
+                                Access::StructMember(ty, index as u32),
+                                |access_stack| {
+                                    self.write_workgroup_variable_initialization(
+                                        module,
+                                        member.ty,
+                                        access_stack,
+                                        level,
+                                    )
+                                },
+                            )?;
+                        }
+                    }
+                    _ => unreachable!(),
+                }
+            } else {
+                write!(self.out, "{}", level)?;
+                access_stack.write(&mut self.out, &self.names)?;
+                writeln!(self.out, " = {{}};")?;
+            }
+
+            Ok(())
+        }
+    }
+
+    // TODO: we do extra work here, replace with a check of TypeFlags::CONSTRUCTIBLE
```

Comment:
I'd like some thoughts on this. Currently don't retain the `TypeFlags` after doing the validation.